### PR TITLE
Wild docs appear!

### LINKED ISSUE #11: Fill and generate docs
ATM I'm using [this docs format](https://pythonhosted.org/an_example_pypi_project/sphinx.html) for things which may be hard to understand without documenting.
The best would be if we fill all missing docs and add some github action to generate page from docstrings.
Github pages?

### LINKED ISSUE #25: Write docstrings for everything
Also fix all of the darglint errors. Darglint throws error codes which can be checked there: https://github.com/terrencepreilly/darglint#error-codes

### PR DESCRIPTION
**Pull request by JuniorJPDJ**
^__^

### CODE REVIEW

File: music_service_async_interface/__init__.py [hunk 1]
```diff
@@ -77,6 +77,13 @@ def __lt__(self, other):
 
 
 class Session(ABC):
+    """Main abstract class holding session and other data neccessary to interact with music service
+
+    Used to store auth data and is start point of library.
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: Starting point of library; used to store auth data

File: music_service_async_interface/__init__.py [hunk 2]
```diff
@@ -77,6 +77,13 @@ def __lt__(self, other):
...
+    Class inheriting from this class should fill `_obj` and `_quality` with own classes
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: s/Class/Classes/
s/own/their own/

File: music_service_async_interface/__init__.py [hunk 3]
```diff
@@ -77,6 +77,13 @@ def __lt__(self, other):
...
+    extending :class:`Object` and :class:`AudioQuality`.
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: Add an example

File: music_service_async_interface/__init__.py [hunk 4]
```diff
@@ -94,6 +101,11 @@ def __init__(self, sess: Optional[aiohttp.ClientSession] = None):
 
     @property
     def preferred_audio_quality(self) -> AudioQuality:
+        """Preferred (maximum you want to get) :class:`AudioQuality` for tracks downloaded using this session
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: either remove the parenthesis or call it an `upper limit`

File: music_service_async_interface/__init__.py [hunk 5]
```diff
@@ -104,6 +116,11 @@ def preferred_audio_quality(self, quality: AudioQuality):
 
     @property
     def required_audio_quality(self) -> AudioQuality:
+        """Required (minimal accepted) :class:`AudioQuality` for tracks downloaded using this session
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: same as above but `lower limit`

File: music_service_async_interface/__init__.py [hunk 6]
```diff
@@ -113,6 +130,17 @@ def required_audio_quality(self, quality: AudioQuality):
         self._required_audio_quality = quality
 
     async def object_from_url(self, url: str) -> "Object":
+        """Gets object from music service based on URL
+        Autodetects URL type!
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: Can be either URL or String

**Review by loboda4450 — APPROVED** [review]

File: music_service_async_interface/__init__.py [hunk 7]
```diff
@@ -147,27 +198,58 @@ async def parse_urls(self, long_string: str) -> AsyncGenerator["Object", None]:
...
         """
         raise NotImplementedError
 
     @classmethod
     @abstractmethod
     async def from_id(cls, sess: Session, id_) -> "Object":
+        """Fetches corresponding object from music service based on ID
+
+        :param sess: :class:`Session` instance to use when loading data from music service
+        :param id_: music service ID of :class:`Object`
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: Is it id of a track, a playlist? I don't quite get it. Add an example. Remove 'music service'

File: music_service_async_interface/__init__.py [hunk 8]
```diff
@@ -147,27 +198,58 @@ async def parse_urls(self, long_string: str) -> AsyncGenerator["Object", None]:
...
+
+        :return: URL to object
+        """
         ...
 
     @abstractmethod
     def get_id(self):
+        """Gets object's music service ID
+
+        :return: music service ID of object
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: what's with the music service id? What is it?

File: music_service_async_interface/__init__.py [hunk 9]
```diff
@@ -176,35 +258,55 @@ class Cover(ABC):
 
     @abstractmethod
     def get_url(self, *args, **kwargs) -> str:
+        """Gets :class:`Cover` image URL
+
+        :param args: allows subfunction to define custom positional arguments
+        :param kwargs: allows subfunction to define custom keyword arguments
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: Can you describe it better to me? What is a subfunction?

File: music_service_async_interface/__init__.py [hunk 10]
```diff
@@ -215,41 +317,162 @@ async def get_file_url(
...
-        :param required_quality: Quality requirement for track
-        :param preferred_quality: Preferred audio quality, which we are asking for
-        :raise: :class:`InsufficientAudioQuality` when quality requirements are not met
-        :return: Music file URL
+        """Fetches direct URL to music file from music service
+
+        :param required_quality: required (minimal accepted) :class:`AudioQuality` for track
+        if ommited value from session is used
+        :param preferred_quality: preferred (maximum) :class:`AudioQuality` you want to get
+        if ommited value from session is used
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: Same changes as above

File: music_service_async_interface/__init__.py [hunk 11]
```diff
@@ -215,41 +317,162 @@ async def get_file_url(
...
 
 
 class ObjectCollection:
+    """Base class for any collections of objects
+
+    Makes sure particular method for object type is overridden.
+    Allows easy iteration over all object types in collection.
+
+    wrong example:
+    >>> class Album(ObjectCollection[Track]):
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: I'm not sure we should include anti-patterns and wrong examples. People often copy the examples blindly and just hope for them to work. Also it's easier to understand when there are working examples everywhere. We could simply say that the `ObjectCollection[Track]` class has to have the `tracks()` method.

File: music_service_async_interface/__init__.py [hunk 12]
```diff
@@ -215,41 +317,162 @@ async def get_file_url(
...
+    ...
+
+    You cannot create `ObjectCollection[Track]` subclass without method `tracks()`.
+    Way of creating method names are defined by `plural_noun(obj_name)` function.
+
+    >>> a = Album()
+    Traceback (most recent call last):
+      File "<ipython-input-16-a40ab37647e0>", line 1, in <module>
+        a = Album()
+    TypeError: Can't instantiate abstract class Album with abstract method tracks
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: We probably should make it raise `TypeError: Can't instantiate abstract class Album with 'tracks' method unimplemented.`

**Review by wvffle — CHANGES_REQUESTED** [review]

File: music_service_async_interface/__init__.py (hunk 3 above)
Comment (JuniorJPDJ) [reply music_service_async_interface/__init__.py]: there's one

File: music_service_async_interface/__init__.py (hunk 7 above)
Comment (JuniorJPDJ) [reply music_service_async_interface/__init__.py]: any of them, just ID of object in music service id-space, object can be playlist, track, whatever
it depends on music service, that's why I mentioned `music service` there. Please remember that's a generic.

File: music_service_async_interface/__init__.py [hunk 13]
```diff
@@ -104,6 +131,11 @@ def preferred_audio_quality(self, quality: AudioQuality):
 
     @property
     def required_audio_quality(self) -> AudioQuality:
+        """Required (minimal accepted) :class:`AudioQuality` for tracks downloaded using this session
+        Can be overridden when downloading track, this is default value.
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: is a

File: music_service_async_interface/__init__.py (hunk 6 above)
Comment (JuniorJPDJ) [reply music_service_async_interface/__init__.py]: what do you mean?
anyway - it should be `fetches`, wtf

File: music_service_async_interface/__init__.py [hunk 14]
```diff
@@ -94,6 +116,11 @@ def __init__(self, sess: Optional[aiohttp.ClientSession] = None):
 
     @property
     def preferred_audio_quality(self) -> AudioQuality:
+        """Preferred (maximum you want to get) :class:`AudioQuality` for tracks downloaded using this session
+        Can be overridden when downloading track, this is default value.
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: is a

File: music_service_async_interface/__init__.py (hunk 1 above)
Comment (JuniorJPDJ) [reply music_service_async_interface/__init__.py]: \+
should be 2 sentences

File: music_service_async_interface/__init__.py (hunk 6 above)
Comment (wvffle) [reply music_service_async_interface/__init__.py]: s/Autodetects URL type!/Can be either URL or String/

File: music_service_async_interface/__init__.py (hunk 8 above)
Comment (JuniorJPDJ) [reply music_service_async_interface/__init__.py]: same as above

File: music_service_async_interface/__init__.py (hunk 9 above)
Comment (JuniorJPDJ) [reply music_service_async_interface/__init__.py]: I was forced to do this docs by linter, those are useless.
Subfunction is function overriding this function, but if you know better fitting name - tell me.

File: music_service_async_interface/__init__.py (hunk 12 above)
Comment (JuniorJPDJ) [reply music_service_async_interface/__init__.py]: That's not my error, it's error from python ABS.
It's not obvious, that's why this example is here.

File: music_service_async_interface/__init__.py (hunk 7 above)
Comment (wvffle) [reply music_service_async_interface/__init__.py]: then it's phrased badly in my opinion. Something like `ID of Object provided by the music service` would be better I presume

File: music_service_async_interface/__init__.py (hunk 12 above)
Comment (wvffle) [reply music_service_async_interface/__init__.py]: ok.

File: music_service_async_interface/__init__.py (hunk 11 above)
Comment (JuniorJPDJ) [reply music_service_async_interface/__init__.py]: explanation below

File: music_service_async_interface/__init__.py (hunk 9 above)
Comment (wvffle) [reply music_service_async_interface/__init__.py]: `overriding function`?

**Review by wvffle — CHANGES_REQUESTED** [review]

File: music_service_async_interface/__init__.py (hunk 9 above)
Comment (wvffle) [reply music_service_async_interface/__init__.py]: `implemented function`?

File: music_service_async_interface/__init__.py (hunk 6 above)
Comment (JuniorJPDJ) [reply music_service_async_interface/__init__.py]: Not the point. Gonna mumble it ;V

Comment (JuniorJPDJ) [reply music_service_async_interface/__init__.py]: K, consensus met!

File: music_service_async_interface/__init__.py (hunk 11 above)
Comment (JuniorJPDJ) [reply music_service_async_interface/__init__.py]: k, we met some consensus

**Review by wvffle — APPROVED** [review]

**Pull request merged**